### ISSUE TITLE
torch.clamp_ not inplace during backward

### ISSUE THREAD

**Issue by chengmengli06** [pytorch/pytorch#33373]
I use torch.clamp_ to avoid forward overflow for fp16 calculation. 
It does not consume extra memory during forward pass. 
However, it has the **same gpu memory cost** as torch.clamp during backward pass. 
In comparison, torch.relu_ does not take extra gpu memory during both forward and backward pass.

torch version: 1.3.0
torch.version.cuda: 10.1.243

```
_conv3d = nn.Conv3d
class Conv3dWrapper(_conv3d):
  def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0, dilation=1,
               groups=1, bias=True, padding_mode='zeros'):
    super(Conv3dWrapper, self).__init__(in_channels, out_channels, kernel_size, stride,
                             padding, dilation, groups, bias, padding_mode)

  def forward(self, input):
    return torch.clamp_(super(Conv3dWrapper, self).forward(input), -65504, 65504)

_conv_trans3d = nn.ConvTranspose3d
class ConvTranspose3dWrapper(_conv_trans3d):
  def __init__(self, in_channels, out_channels, kernel_size, stride=1, padding=0,
            output_padding=0, groups=1, bias=True, dilation=1, padding_mode='zeros'):
    super(ConvTranspose3dWrapper, self).__init__(in_channels, out_channels, kernel_size,
            stride, padding, output_padding, groups, bias, dilation, padding_mode)

  def forward(self, input):
    return torch.clamp_(super(ConvTranspose3dWrapper, self).forward(input), -65504, 65504)

```

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by VitalyFedyunin** [comment]
Indeed, looks like there is no inplace backward for clamp_ in our code. Feel free to send us PR implementing it.

**Comment by albanD** [comment]
Hi,

This is expected I'm afraid.
Since the gradient buffers are re-used during the backward pass by the engine, inplace modifications of the gradient is **never** allowed (from backward Nodes or backward hooks).

You will need to write a custom autograd.Function that wraps all these ops together to avoid this copy (make sure your backward don't change the given Tensor inplace!).

**Comment by VitalyFedyunin** [comment]
I think I must correct myself here. You cannot inplace modify gradient, but you can send us pr, which will modify `clamp_backward` and fuse all operations. This will prevent pytorch from using additional memory for temporary tensors.

**Comment by albanD** [comment]
Ho I did not check the `clamp_backward` implementation. It can indeed be slightly optimized, and we would be happy to accept a PR for that.

**Comment by chengmengli06** [comment]
I think for torch.clamp_, the gradient could be modified inplace, but for torch.clamp, the gradient should not be modified inplace. Is my understanding correct? @albanD @VitalyFedyunin

**Comment by chengmengli06** [comment]
I cannot understand why I cannot modify gradient inplace, where are the gradients used in the engine? Is my following AutoGrad implementation correct? @albanD @VitalyFedyunin 
```
class MyClamp(torch.autograd.Function):
  @staticmethod
  def forward(ctx, input):
    input = torch.clamp_(input, -65504, 65504)
    ctx.save_for_backward(input)
    return input

  @staticmethod
  def backward(ctx, grad_output):
    input, = ctx.saved_tensors
    #grad_output *= (((input < 65504)  | (grad_output > 0)) &\
    #                ((input > -65504) | (grad_output < 0)))
    grad_output *= ((input < 65504) & (input > -65504))
    return grad_output
```

**Comment by albanD** [comment]
Hi,

Because depending on what are the other operations around this particular Function, this `grad_output` can be used for hooks, for setting the `.grad` of some Tensors or as grad_output for other Functions.

In this example, the user-provided grad_output is modified inplace for example:
```python
import torch

class MyFn(torch.autograd.Function):
  @staticmethod
  def forward(ctx, input):
    return input.clone()

  @staticmethod
  def backward(ctx, grad_output):
    grad_output.fill_(42)
    return grad_output


a = torch.rand(10, requires_grad=True)

loss = MyFn.apply(a).sum()

grad_loss = torch.ones([])
print("grad_loss is:", grad_loss)
loss.backward(grad_loss)

print(a.grad)
print("grad_loss is:", grad_loss)
```
